File: pipeline/migrate_slugs.js
// One-time migration: add slug column, backfill, rename files to title-based names, fix paths.
// Usage: node --experimental-sqlite pipeline/migrate_slugs.js
'use strict';
const fs = require('node:fs');
const path = require('node:path');
const { open, ROOT } = require('./lib/db');
const { uniqueSlug } = require('./lib/slug');

const oldFileId = (id) => id.replace(/[^a-z0-9._-]/gi, '_').slice(0, 180);
const mv = (a, b) => { if (fs.existsSync(a) && a !== b) { fs.mkdirSync(path.dirname(b), { recursive: true }); fs.renameSync(a, b); return true; } return false; };

function main() {
  const db = open();
  const cols = db.prepare('PRAGMA table_info(papers)').all().map((c) => c.name);
  if (!cols.includes('slug')) db.exec('ALTER TABLE papers ADD COLUMN slug TEXT');

  const papers = db.prepare('SELECT id, title, slug, pdf_path, text_path FROM papers ORDER BY discovered_at').all();
  let renamedFiles = 0, sluggedRows = 0;

  for (const p of papers) {
    const slug = p.slug || uniqueSlug(db, p.title, p.id);
    if (!p.slug) { db.prepare('UPDATE papers SET slug=? WHERE id=?').run(slug, p.id); sluggedRows++; }

    const oldBase = oldFileId(p.id);
    // pdf
    if (mv(path.join(ROOT, 'store/pdfs', oldBase + '.pdf'), path.join(ROOT, 'store/pdfs', slug + '.pdf'))) renamedFiles++;
    // text
    if (mv(path.join(ROOT, 'store/text', oldBase + '.txt'), path.join(ROOT, 'store/text', slug + '.txt'))) renamedFiles++;
    // summary dir
    if (mv(path.join(ROOT, 'store/summaries', oldBase), path.join(ROOT, 'store/summaries', slug))) renamedFiles++;

    // fix DB paths
    const newPdf = path.join('store/pdfs', slug + '.pdf');
    const newTxt = path.join('store/text', slug + '.txt');
    if (p.pdf_path) db.prepare('UPDATE papers SET pdf_path=? WHERE id=?').run(newPdf, p.id);
    if (p.text_path) db.prepare('UPDATE papers SET text_path=? WHERE id=?').run(newTxt, p.id);
    // fix summary_versions paths
    for (const sv of db.prepare('SELECT version, path FROM summary_versions WHERE paper_id=?').all(p.id)) {
      const fn = path.basename(sv.path);
      db.prepare('UPDATE summary_versions SET path=? WHERE paper_id=? AND version=?')
        .run(path.join('store/summaries', slug, fn), p.id, sv.version);
    }
  }
  db.close();
  console.log(`migrated: ${sluggedRows} rows slugged, ${renamedFiles} files/dirs renamed`);
}
main();

File: pipeline/tools/migrate_slugs.py
"""One-time migration: add slug column, backfill, rename files to title-based names,
fix DB paths. Idempotent (re-running after migration is a no-op).
Usage: python3 pipeline/tools/migrate_slugs.py
"""
import re
import shutil
from pathlib import Path

# --- path shim: 让 `from lib...` 解析到 pipeline/lib，无论本文件在哪个子目录 ---
import os as _os, sys as _sys
_sys.path.insert(0, _os.path.dirname(_os.path.dirname(_os.path.abspath(__file__))))
from lib.db import open_db, ROOT
from lib.slug import unique_slug


def old_file_id(pid):
    return re.sub(r"[^a-z0-9._-]", "_", pid, flags=re.I)[:180]


def mv(a, b):
    a, b = Path(a), Path(b)
    if a.exists() and a != b:
        b.parent.mkdir(parents=True, exist_ok=True)
        shutil.move(str(a), str(b))
        return True
    return False


def main():
    conn = open_db()
    papers = conn.execute("SELECT id, title, slug, pdf_path, text_path FROM papers ORDER BY discovered_at").fetchall()
    renamed = slugged = 0
    for p in papers:
        slug = p["slug"] or unique_slug(conn, p["title"], p["id"])
        if not p["slug"]:
            conn.execute("UPDATE papers SET slug=? WHERE id=?", (slug, p["id"]))
            slugged += 1
        old = old_file_id(p["id"])
        if mv(ROOT / "store/pdfs" / (old + ".pdf"), ROOT / "store/pdfs" / (slug + ".pdf")):
            renamed += 1
        if mv(ROOT / "store/text" / (old + ".txt"), ROOT / "store/text" / (slug + ".txt")):
            renamed += 1
        if mv(ROOT / "store/summaries" / old, ROOT / "store/summaries" / slug):
            renamed += 1
        if p["pdf_path"]:
            conn.execute("UPDATE papers SET pdf_path=? WHERE id=?", (f"store/pdfs/{slug}.pdf", p["id"]))
        if p["text_path"]:
            conn.execute("UPDATE papers SET text_path=? WHERE id=?", (f"store/text/{slug}.txt", p["id"]))
        for sv in conn.execute("SELECT version, path FROM summary_versions WHERE paper_id=?", (p["id"],)).fetchall():
            fn = Path(sv["path"]).name
            conn.execute("UPDATE summary_versions SET path=? WHERE paper_id=? AND version=?",
                         (f"store/summaries/{slug}/{fn}", p["id"], sv["version"]))
    conn.commit()
    conn.close()
    print(f"migrated: {slugged} rows slugged, {renamed} files/dirs renamed")


if __name__ == "__main__":
    main()
